src/lib/drivers/device/nuttx/SPI.cpp

```diff
@@ -84,87 +78,99 @@ SPI::SPI(const char *name,
 
 SPI::~SPI()
 {
-	// XXX no way to let go of the bus...
+	perf_free(_isr_deferred);
 }
 
 int
 SPI::init()
 {
-	int ret = OK;
-
 	/* attach to the spi bus */
 	if (_dev == nullptr) {
 		_dev = px4_spibus_initialize(get_device_bus());
 	}
 
 	if (_dev == nullptr) {
 		DEVICE_DEBUG("failed to init SPI");
-		ret = -ENOENT;
-		goto out;
+		return PX4_ERROR;
 	}
 
 	/* deselect device to ensure high to low transition of pin select */
 	SPI_SELECT(_dev, _device, false);
 
 	/* call the probe function to check whether the device is present */
-	ret = probe();
-
-	if (ret != OK) {
+	if (probe() != PX4_OK) {
 		DEVICE_DEBUG("probe failed");
-		goto out;
+		return PX4_ERROR;
 	}
 
 	/* do base class init, which will create the device node, etc. */
-	ret = CDev::init();
-
-	if (ret != OK) {
+	if (CDev::init() != PX4_OK) {
 		DEVICE_DEBUG("cdev init failed");
-		goto out;
+		return PX4_ERROR;
 	}
 
 	/* tell the workd where we are */
 	DEVICE_LOG("on SPI bus %d at %d (%u KHz)", get_device_bus(), PX4_SPI_DEV_ID(_device), _frequency / 1000);
 
-out:
-	return ret;
+	return PX4_OK;
+}
+
+void SPI::lock(struct spi_dev_s *dev)
+{
+	_is_locked |= (1 << _device_id.devid_s.bus);
```

Comment:
@dakejahl this needs to be atomic. It's a shared state potentially written to by multiple threads.
You can use the API in https://github.com/PX4/Firmware/pull/11328.

Also this assumes that all `SPI` instances for a given bus run on the same thread. Can you add this as comment here to make the assumption explicit?

Reply:
FYI once we're sure there's no longer a mix of PX4 SPI drivers running out of HRT and threads we can drop all of this and lean on the underlying nuttx SPI_LOCK (semaphore).

Reply:
What I meant to emphasize was that this only exists for the case of a potential HRT and thread SPI conflict.

Reply:
> this needs to be atomic. It's a shared state potentially written to by multiple threads.
You can use the API in #11328.

Cool! I'll add that in as soon as #11328 is merged.